### PR TITLE
Muzzle improvements: docs, javadocs, renamings and minor refactoring

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Contents:
* Renamed almost all muzzle classes, moved some of them to two different packages (`collector` or `matcher`) depending on how and when they're used;
* Added/improved javadocs in classes that had no/poor documentation;
* Documented compile-time and runtime muzzle functionality in `muzzle.md`;
* Documented muzzle gradle plugin usage (as much as I understand it);
* Split `Reference.Flag` to several different enums;
* Refactored `HelperReferenceWrapper` so that it uses Java 8 streams instead of Guava.

This PR does not solve #536, but at least it clears up some of the questions that appeared there.

### CODE REVIEW

**Review by trask — COMMENTED** [review]
Great stuff!

> * 👍 Renamed almost all muzzle classes, moved some of them to two different packages (`collector` or `matcher`) depending on how and when they're used;
> * ❤️ Added/improved javadocs in classes that had no/poor documentation;
> * ❤️ Documented compile-time and runtime muzzle functionality in `muzzle.md`;
> * ❤️ Documented muzzle gradle plugin usage (as much as I understand it);
> * 👍 Split `Reference.Flag` to several different enums;
> * 👍 Refactored `HelperReferenceWrapper` so that it uses Java 8 streams instead of Guava.

File: docs/contributing/muzzle.md [hunk 1]
```diff
@@ -1,42 +1,107 @@
...
+to verify that the class loader of the instrumented type has all necessary symbols (classes,
+methods, fields). If the `ReferenceMatcher` finds any mismatch between collected references and the
+actual application classpath types the whole instrumentation is discarded.
+
+It is worth noting that because the muzzle check is expensive, it is only performed after a match
+has been made by the `InstrumenterDefault#classLoaderMatcher()` and `Instrumenter.Default#typeMatcher()`
+matchers.
+
+The source code of the runtime muzzle matcher is located in the `javaagent-tooling` module,
+class `Instrumenter.Default` and package `io.opentelemetry.javaagent.tooling.muzzle.matcher`.
```
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply docs/contributing/muzzle.md:49]: I think I'll leave it here for the time being - it seems that placing it inside `Instrumenter.Default` was purposeful, see the `Optimization` comment at top of the `MuzzleMatcher#matches()` method.

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/collector/MuzzleCodeGenerator.java [hunk 2]
```diff
@@ -462,9 +465,21 @@ public void visitEnd() {
...
-    public class InitializeFieldVisitor extends MethodVisitor {
-      public InitializeFieldVisitor(MethodVisitor methodVisitor) {
+    private static final Pattern ANONYMOUS_ENUM_CONSTANT_CLASS =
+        Pattern.compile("(?<enumClass>.*)\\$[0-9]+$");
+
+    // drops "$1" suffix for enum constants that override/implement super class methods
+    private String getEnumClassInternalName(Reference.Flag flag) {
+      String fullInternalName = Utils.getInternalName(flag.getClass().getName());
+      Matcher m = ANONYMOUS_ENUM_CONSTANT_CLASS.matcher(fullInternalName);
+      return m.matches() ? m.group("enumClass") : fullInternalName;
```
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/collector/MuzzleCodeGenerator.java:475]: It's a really awesome feature when you're working with large patterns with many groups - although I use them pretty much everywhere, just for the readability benefit 😄

File: docs/contributing/muzzle.md [hunk 3]
```diff
@@ -1,42 +1,107 @@
...
+it's not an optional feature.
+
+The gradle plugin defines two tasks:
+
+* `printReferences` task prints all API references in a given module:
+    ```sh
+    ./gradlew :instrumentation:google-http-client-1.19:printReferences
+    ```
+
+* `muzzle` task runs the runtime muzzle verification against different library versions:
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply docs/contributing/muzzle.md:69]: not related to this PR, but should we rename the `printReferences` to something that includes muzzle in the task name? e.g. `muzzlePrintReferences`?

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/collector/ReferenceCollector.java [hunk 4]
```diff
@@ -110,14 +122,14 @@ private static String internalPackageName(String internalName) {
    *
    * @return A reference flag with the required level of access.
    */
-  private static Reference.Flag computeMinimumClassAccess(Type from, Type to) {
+  private static Flag computeMinimumClassAccess(Type from, Type to) {
```
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/collector/ReferenceCollector.java]: Done.

**Review by trask — APPROVED** [review]
🎉

File: docs/contributing/muzzle.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (trask) [reply docs/contributing/muzzle.md]: @pavolloffay makes sense 👍

**Pull request merged**